mlir/lib/Dialect/XeGPU/IR/XeGPUDialect.cpp

```diff
@@ -33,6 +37,55 @@ void XeGPUDialect::initialize() {
       >();
 }
 
+/// Generates instructions to compute offsets for a subgroup identified by
+/// its multidimensional indices (sgId), using the specified subgroup layout
+/// (sgLayout), subgroup data dimensions (sgShape), and the overall data
+/// dimensions (shape).
+static SmallVector<SmallVector<Value>>
+genOffsetsComputations(OpBuilder &builder, Location loc,
+                       SmallVector<Value> sgId, ArrayRef<int64_t> sgLayout,
+                       ArrayRef<int64_t> sgShape, ArrayRef<int64_t> shape) {
```

Comment:
consider generalizing sgShape to sgDataStride. 
From the caller size, need to set as following : 
sgDataStride[i] == sgShape[i] if (wgData[i] % (sgData[i]*sgLayout[i])) == 0
sgDataStride[i] == 0  if wgData[i] == sgData[i]